File: sdk/ctv/tvos/Tests/CTVSDKTests/AuctionClientTests.swift
import XCTest
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif
@testable import CTVSDK

final class AuctionClientTests: XCTestCase {
    private var session: URLSession!
    private var config: SDKConfig!

    override func setUp() {
        super.setUp()
        let cfg = URLSessionConfiguration.ephemeral
        cfg.protocolClasses = [MockURLProtocol.self]
        session = URLSession(configuration: cfg)
        config = SDKConfig(appId: "app-1", apiBaseUrl: "http://localhost:4000/api/v1")
    }

    override func tearDown() {
        session.invalidateAndCancel()
        session = nil
        super.tearDown()
    }

    func testNoFill204() {
        MockURLProtocol.requestHandler = { request in
            let resp = HTTPURLResponse(url: request.url!, statusCode: 204, httpVersion: nil, headerFields: nil)!
            return (resp, Data())
        }
        let client = AuctionClient(config: config, session: session)
        let exp = expectation(description: "no_fill")
        client.requestBid(placementId: "p1", adFormat: "interstitial", consent: nil) { result in
            XCTAssertTrue(result.noFill)
            XCTAssertEqual(result.error, .noFill)
            exp.fulfill()
        }
        wait(for: [exp], timeout: 2.0)
    }

    func testParsesEnvelope200() {
        let obj: [String: Any] = [
            "response": [
                "requestId": "r1",
                "bidId": "b1",
                "adapter": "admob",
                "cpm": 1.23,
                "currency": "USD",
                "ttlSeconds": 300,
                "creativeUrl": "https://cdn.test/video.mp4",
                "tracking": ["impression": "http://i", "click": "http://c"],
                "payload": [:]
            ]
        ]
        let data = try! JSONSerialization.data(withJSONObject: obj)
        MockURLProtocol.requestHandler = { request in
            let resp = HTTPURLResponse(url: request.url!, statusCode: 200, httpVersion: nil, headerFields: ["Content-Type":"application/json"])!
            return (resp, data)
        }
        let client = AuctionClient(config: config, session: session)
        let exp = expectation(description: "ok")
        client.requestBid(placementId: "p1", adFormat: "interstitial", consent: nil) { result in
            XCTAssertFalse(result.noFill)
            XCTAssertNil(result.error)
            XCTAssertNotNil(result.win)
            exp.fulfill()
        }
        wait(for: [exp], timeout: 2.0)
    }
}

// MARK: - URLProtocol mock
final class MockURLProtocol: URLProtocol {
    static var requestHandler: ((URLRequest) throws -> (HTTPURLResponse, Data))?
    override class func canInit(with request: URLRequest) -> Bool { return true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { return request }
    override func startLoading() {
        guard let handler = MockURLProtocol.requestHandler else { fatalError("No handler set") }
        do {
            let (response, data) = try handler(request)
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
        } catch {
            client?.urlProtocol(self, didFailWithError: error)
        }
    }
    override func stopLoading() {}
}

File: sdk/ctv/android-tv/src/test/kotlin/network/AuctionClientTests.kt
package network

import android.content.Context
import com.google.gson.Gson
import com.rivalapexmediation.ctv.SDKConfig
import com.rivalapexmediation.ctv.network.AuctionClient
import com.rivalapexmediation.ctv.network.LoadError
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class AuctionClientTests {
    private lateinit var server: MockWebServer
    private lateinit var context: Context

    @Before
    fun setUp() {
        server = MockWebServer()
        server.start()
        context = RuntimeEnvironment.getApplication()
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `returns no_fill on 204`() {
        server.enqueue(MockResponse().setResponseCode(204))
        val cfg = SDKConfig(appId = "app-1", apiBaseUrl = server.url("/api/v1").toString())
        val client = AuctionClient(context, cfg)
        var err: LoadError? = null
        client.requestBid("p1", "interstitial", 0.0, null) { res ->
            err = res.error
        }
        Thread.sleep(100) // wait async
        assertTrue(err is LoadError.NoFill)
    }

    @Test
    fun `parses 200 success envelope`() {
        val body = mapOf(
            "response" to mapOf(
                "requestId" to "r1",
                "bidId" to "b1",
                "adapter" to "admob",
                "cpm" to 1.23,
                "currency" to "USD",
                "ttlSeconds" to 300,
                "creativeUrl" to "https://cdn.test/video.mp4",
                "tracking" to mapOf("impression" to "http://i", "click" to "http://c"),
                "payload" to emptyMap<String, Any>()
            )
        )
        server.enqueue(MockResponse().setResponseCode(200).setBody(Gson().toJson(body)).setHeader("Content-Type","application/json"))
        val cfg = SDKConfig(appId = "app-1", apiBaseUrl = server.url("/api/v1").toString())
        val client = AuctionClient(context, cfg)
        var ok = false
        var err: LoadError? = null
        client.requestBid("p1", "interstitial", 0.0, null) { res ->
            ok = res.win != null && res.error == null
            err = res.error
        }
        Thread.sleep(100)
        assertTrue(ok)
        assertEquals(null, err)
    }
}
